packages/@aws-cdk/aws-stepfunctions-tasks/lib/emr-create-cluster.ts

```diff
@@ -0,0 +1,107 @@
+import iam = require('@aws-cdk/aws-iam');
+import sfn = require('@aws-cdk/aws-stepfunctions');
+import { Stack } from '@aws-cdk/core';
+import { getResourceArn } from './resource-arn-suffix';
+
+/**
+ * Properties for EmrCreateCluster
+ */
+export interface EmrCreateClusterProps {
+  /**
+   * The JSON that you want to provide to your CreateCluster call as input.
+   *
+   * This uses the same syntax as the RunJobFlow API.
+   *
+   * @see https://docs.aws.amazon.com/emr/latest/APIReference/API_RunJobFlow.html
+   */
+  readonly clusterConfiguration: sfn.TaskInput;
```

Comment:
Personally, I think I would prefer all the fields of `clusterConfiguration` to be inlined here. I know this is a lot more work for you as a construct writer, but it adds the autocomplete and type safety that is one of the big benefits the CDK provides.

Reply:
I updated the emr-add-step task with strongly typed input parameters rather than the `TaskInput`. I have a first pass at doing the same for the emr-create-cluster in a branch of my fork linked below.

In that branch, I only added strong typing for the top-level cluster configuration parameters, taking inputs and requirements from the RunJobFlow API. I used L2 Constructs for the input Roles, and renamed the JobFlowRole to ClusterRole to coincide with similar renames in the EMR/Step Functions Integration.  

I didn't implement all levels of the cluster configuration parameters, as this would be duplicating the generated `aws-emr.CfnClusterProps` and associated nested interfaces. The requirements of those nested parameters are dependent on the EMR Release Label requested at launch and are validated by the API. Attempting to implement that fully on the Step Functions Task seems like the wrong place... 

Link to the branch with my first pass:
https://github.com/chamcca/aws-cdk/blob/explicit-create-cluster-params/packages/%40aws-cdk/aws-stepfunctions-tasks/lib/emr-create-cluster.ts

Reply:
I'm not sure about the duplication. Can you point me to specific instances of duplication that you're unsure about?

As another note, in the CDK we tend to generate IAM roles for you instead of requiring you to do this yourself. Any chance that is a pattern you can take on as well?

At first glance, what you did for `AddStep` is exactly what I had in mind. CDK wants to be a complete abstraction for AWS, we don't stop at saying "and this is where you put in your API parameters, go figure out what those are", we help you all the way, removing the need to cross-ref other docs as much as possible.

> The requirements of those nested parameters are dependent on the EMR Release Label requested at launch and are validated by the API

Unfortunately I don't know what that means.

Reply:
A deep implementation of the cluster configuration parameters becomes a recreation of the CfnClusterProps and other nested interfaces that are generated in https://github.com/aws/aws-cdk/tree/master/packages/%40aws-cdk/aws-emr/lib in **emr.generated.ts**.

I'm looking into whether it's possible to use some of the generated interfaces for the nested parameters. For example, using the `CfnCluster.JobFlowInstancesConfigProperty` and related nested interfaces from **emr.generated.ts** as the `instances` parameter on the `EmrCreateCluster` task. If this works, is use of these Cfn escape hatches acceptable? Or is a deep recreation of those interfaces required? There are a lot of nested interfaces.

The validation of some of the nested parameters is dependent on the value of the top-level `emrReleaseLabel` version (e.g. 'emr-5.20.0' vs 'emr-5.28.0'). The EMR API performs the validation checks. I don't think it's feasible or wise to try to perform those validations in the `EmrCreateCluster` task.

Reply:
Also, what do you think of making the Roles optional and exposing them as properties of the `EmrCreateCluster` task? If not provided to the constructor the task will generate them. We frequently see customers creating these Roles, setting up access policies, then reusing them across fleets of transient and/or persistent clusters.

Reply:
Yep, that seems like a good thing to do.

Reply:
> For example, using the CfnCluster.JobFlowInstancesConfigProperty

No, and I'll tell you why not: we want to keep the L1 and L2 layers separate, so that we can provide interface stability on the L2 even if the L1 changes.

The ideal solution would have been to make an L2 mirror of  `CfnCluster.JobFlowInstancesConfigProperty` (let's say it's called `JobFlowInstances`, and you would re-use or extend that here.

That is still a deep mirror of all the properties, but at least it's not 2 deep mirrors.

> making the Roles optional and exposing them as properties 

Yes, that seems like a good idea.

Reply:
@rix0rrr I've pushed a large update to this PR that implements most of the requested changes to EmrCreateCluster. It implements `Instances.InstanceFleets` but not the `Instances.InstanceGroups`, I'm still working on those and running through deployment tests for both Fleets and Groups. Can you look over what I've pushed so far? I'll make any necessary changes while continuing to work on the InstanceGroups.

Reply:
Also, Roles are now optional and created if not provided.

Reply:
Full implementation of the `Instances.InstanceFleets` and `Instances.InstanceGroups` has been pushed. I am running additional deployment tests, but these take some time. So far everything looks good.